### PR TITLE
Teradata Vantage - Added SSL support

### PR DESCRIPTION
**Pull request by sc250072**
Modified database size from 1GB to 120 MB for schema creating from Airbyte.

## What
*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*

## How
*Describe the solution*

## Recommended reading order
1. `[TeradataDestination.java](https://github.com/airbytehq/airbyte/compare/master...Teradata:airbyte:Teradata_SSL_Support?expand=1#diff-81950dec5336116d6abe1fb35b027346cb6538ed93f6d8f727b559f2fae1af46)`
2. `airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataSqlOperations.java`
3. 'airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json'

## 🚨 User Impact 🚨
No breaking changes. User can load data from any airbyte supported source to teradata through airbyte.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

SSL_UnitResult

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

SSL_AcceptanceResult

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java [hunk 1]
```diff
@@ -12,57 +12,108 @@
...
-    if (config.has(JdbcUtils.JDBC_URL_PARAMS_KEY)) {
-      configBuilder.put(JdbcUtils.JDBC_URL_PARAMS_KEY, config.get(JdbcUtils.JDBC_URL_PARAMS_KEY).asText());
+    public TeradataDestination() {
+        super(DRIVER_CLASS, new ExtendedNameTransformer(), new TeradataSqlOperations());
+    }
+
+    @Override
+    protected Map<String, String> getDefaultConnectionProperties(final JsonNode config) {
+        final Map<String, String> additionalParameters = new HashMap<>();
+        LOGGER.info("teradata config obj : " + config);
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java:56]: please remove this log statement to avoid leaking any sensitive config information

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  @Override
+  protected JsonNode getConfig() {
+    return configJson;
+  }
+
+  public JsonNode getStaticConfig() throws Exception {
+    final JsonNode config = Jsons.deserialize(Files.readString(Paths.get("secrets/sslconfig.json")));
+    ((ObjectNode) config).put("ssl", true);
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:59]: why not have `ssl` and `sslmode` specified in `secrets/sslconfig.json` instead of adding them programmatically.

File: airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json [hunk 3]
```diff
@@ -21,28 +21,153 @@
...
-        "order": 5
+        "order": 2
       },
       "schema": {
         "title": "Default Schema",
         "description": "The default schema tables are written to if the source does not specify a namespace. The usual value for this field is \"public\".",
         "type": "string",
-        "examples": ["public"],
-        "default": "public",
+        "examples": ["airbyte_td"],
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json:37]: this change seems unnecessary

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java (hunk 2 above)
Comment (grishick) [reply airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:59]: also, `spec.json` expects `ssl_mode` but here you are adding `sslmode`

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import java.util.Map;
+import java.util.Objects;
+import javax.sql.DataSource;
+import org.junit.jupiter.api.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TeradataDestinationSSLAcceptanceTest extends JdbcDestinationAcceptanceTest {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestinationAcceptanceTest.class);
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:38]: this member is not used, please remove

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationAcceptanceTest.java [hunk 5]
```diff
@@ -95,7 +95,7 @@ private List<JsonNode> retrieveRecordsFromTable(final String tableName, final St
   protected void setup(TestDestinationEnv testEnv) {
     final String schemaName = Strings.addRandomSuffix("integration_test_teradata", "_", 5);
     final String createSchemaQuery = String
-        .format(String.format("CREATE DATABASE \"%s\" AS PERM = 1e9 SKEW = 10 PERCENT", schemaName));
+        .format(String.format("CREATE DATABASE \"%s\" AS PERMANENT = 60e6, SPOOL = 60e6 SKEW = 10 PERCENT", schemaName));
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationAcceptanceTest.java:98]: why is this change relevant to this PR?

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  private final ExtendedNameTransformer namingResolver = new ExtendedNameTransformer();
+
+  private JsonNode configJson;
+  private JdbcDatabase database;
+  private DataSource dataSource;
+  private TeradataDestination destination = new TeradataDestination();
+  private final JdbcSourceOperations sourceOperations = JdbcUtils.getDefaultSourceOperations();
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:45]: `sourceOperations` is not used anywhere in this class, so it should be removed

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationAcceptanceTest.java (hunk 5 above)
Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationAcceptanceTest.java:98]: 1GB is not required for running integration tests. With this huge memory setting for each database creating while running integration tests, space issues are coming and test cases getting failed. So changed this to limited memory to make integration tests gets success.

File: airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json (hunk 3 above)
Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/main/resources/spec.json:37]: public is keyword in Teradata system, and it will not allow to create schema with public as default value.  Changed it to airbyte_td

File: airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java (hunk 1 above)
Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java]: Removed this log statement.

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java (hunk 2 above)
Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java]: Removed providing ssl options from this class and will provide latest config json file.

Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java]: Yes, spec.json expects ssl_mode. This testcase was got succeeded when ssl is true, if no ssl_mode specified with any value like require, prefer.., it is taking require as default value. 
image

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java (hunk 4 above)
Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java]: Removed.

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java (hunk 6 above)
Comment (SatishChGit) [reply airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java]: Removed.

**Comment by grishick** [comment]
This PR is looking better. Noting the issues we've discussed over slack to keep track:
 - need to make DB storage footprint configurable via spec instead of hardcoded
 - need to drop test DBs in `tearDown` method in `TeradataDestinationAcceptanceTest`
 - integration tests should be conservative when specifying storage footprint for test DBs (this would help avoid running out of space on the test instance)
 - avoid modifying static config in test code if possible and use config files instead. I.e. `sslconfig.json` should have SSL config enabled

**Comment by SatishChGit** [comment]
> This PR is looking better. Noting the issues we've discussed over slack to keep track:
> 
> * need to make DB storage footprint configurable via spec instead of hardcoded
   I tried to make use of spec to specify storage footprint but couple of testcases are failing. Will work on this in different PR request as currently taking time to investigate on these failures. Will create JIRA from ourside and work in coming sprints. 
> * need to drop test DBs in `tearDown` method in `TeradataDestinationAcceptanceTest`
     Implemented dropping test DBs through tearDown.
> * integration tests should be conservative when specifying storage footprint for test DBs (this would help avoid running out of space on the test instance)
   I tried to make use of spec to specify storage footprint but couple of testcases are failing. Will work on this in different PR request as currently taking time to investigate on these failures. Will create JIRA from ourside and work in coming sprints. 
> * avoid modifying static config in test code if possible and use config files instead. I.e. `sslconfig.json` should have SSL config enabled
 Modified as per this request

**Comment by SatishChGit** [comment]
Integration tests results after all review comments addressed.
image

Unit tests results after all review comments addressed.
image

File: airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+package io.airbyte.integrations.destination.teradata;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TeradataDestinationSSLAcceptanceTest extends TeradataDestinationAcceptanceTest {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestinationSSLAcceptanceTest.class);
+
+  protected  String getConfigFileName() {
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/test-integration/java/io/airbyte/integrations/destination/teradata/TeradataDestinationSSLAcceptanceTest.java:14]: this method is not being called anywhere as far as I can see, which means that this test runs w/o SSL

File: airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java [hunk 8]
```diff
@@ -12,57 +12,107 @@
...
-        config.get(JdbcUtils.HOST_KEY).asText());
-
-    final ImmutableMap.Builder<Object, Object> configBuilder = ImmutableMap.builder()
-        .put(JdbcUtils.USERNAME_KEY, config.get(JdbcUtils.USERNAME_KEY).asText())
-        .put(JdbcUtils.JDBC_URL_KEY, jdbcUrl)
-        .put(JdbcUtils.SCHEMA_KEY, schema);
-
-    if (config.has(JdbcUtils.PASSWORD_KEY)) {
-      configBuilder.put(JdbcUtils.PASSWORD_KEY, config.get(JdbcUtils.PASSWORD_KEY).asText());
+    private static final Logger LOGGER = LoggerFactory.getLogger(TeradataDestination.class);
```
Comment (grishick) [inline airbyte-integrations/connectors/destination-teradata/src/main/java/io/airbyte/integrations/destination/teradata/TeradataDestination.java:27]: LOGGER appears to be unused

**Comment by grishick** [comment]
note: I ran tests locally and they passed

**Comment by grishick** [comment]
@SatishChGit in addition to the two comments I left earlier, could you also rebase your feature branch on latest master or merge latest master into your feature branch - we have some test and instrumentation changes that are not in your feature branch and that's causing test failures.

**Comment by SatishChGit** [comment]
> @SatishChGit in addition to the two comments I left earlier, could you also rebase your feature branch on latest master or merge latest master into your feature branch - we have some test and instrumentation changes that are not in your feature branch and that's causing test failures.

Okay Sure Greg. Will do and run tests

**Comment by grishick** [comment]
I think we can merge this as an alpha connector. Before we can add it to Cloud, we'd need to:
1. integrate spinning up/down of teradata instance into Before/After hooks (we have a discussion in Slack going about this)
2. make sure that only integration tests require a running instance (filed https://github.com/airbytehq/airbyte/issues/23121 to track)
3. some cleanup: https://github.com/airbytehq/airbyte/issues/23122

**Comment by SatishChGit** [comment]
> I think we can merge this as an alpha connector. Before we can add it to Cloud, we'd need to:
> 
> 1. integrate spinning up/down of teradata instance into Before/After hooks (we have a discussion in Slack going about this)
> 2. make sure that only integration tests require a running instance (filed [`TeradataDestinationTest` should not require a running Teradata instance #23121](https://github.com/airbytehq/airbyte/issues/23121) to track)
> 3. some cleanup: [Unused members in `TeradataDestinationTest` #23122](https://github.com/airbytehq/airbyte/issues/23122)

Sure Greg. After merging master to this branch, couple of test cases are failing in my local environment. I am checking on them and will update once these run successfully.

**Review by grishick — APPROVED** [review]
